src/anndata/_core/xarray.py

```diff
@@ -52,13 +53,25 @@ def index(self) -> pd.Index:
         -------
         The index of the of the dataframe as resolved from :attr:`~xarray.Dataset.coords`.
         """
-        coord = get_index_dim(self)
-        return self.indexes[coord]
+        return self.indexes[self.index_dim]
 
     @index.setter
     def index(self, val) -> None:
-        coord = get_index_dim(self)
-        self.coords[coord] = val
+        index_dim = self.index_dim
+        self.coords[index_dim] = (index_dim, val)
+        if isinstance(val, pd.Index) and val.name is not None and val.name != index_dim:
+            self.update(self.rename({self.index_dim: val.name}))
+            del self.coords[index_dim]
+        if "indexing_key" in self.attrs:
+            del self.attrs["indexing_key"]
```

Comment:
The codecov here (and elsewhere) should be addressed

Reply:
I've added a test module for the `Dataset2D` class. The other major piece of new code that is not covered by tests is merge.py:780-786, but that is impossible to test currently as that codepath (`axis==1`) would only be hit when concatenating along columns, which is never the case for obs/var/obsm/varm. I can remove that code and raise an exception there instead, if you think that would be better.